### ISSUE TITLE
[bug] FSDP training does not converge

### ISSUE THREAD

**Issue by pavelgein** [NVIDIA-NeMo/Megatron-Bridge#3826]
### Problem

I am running two similar experiments, the only difference between them is using Megatron-FSDP

There are two plots: red one with FSDP and purle one without  (will attach in the comment)

### Minimal repro

```shell
1. Run Qwen3-8b experiment with and without FSDP
```

### Expected behavior

Losses are the same

### Affected area

area:training

### Regression?

Yes

### Environment

Container: Nemo 26.04
Commit https://github.com/NVIDIA-NeMo/Megatron-Bridge/commit/9437ac7c4890a89e61bf8353367093ba4f46dc84

### Logs

```shell

```

**Comment by pavelgein** [comment]
<img width="680" height="427" alt="Image" src="https://github.com/user-attachments/assets/1d78f0ca-17cb-43b2-80d4-0633da0577e8" />

**Comment by yaoyu-33** [comment]
@pavelgein  thanks for reporting. Can you share your fsdp run script?

**Comment by pavelgein** [comment]
I attached the script file, it misses the dataset section (I was training with context length 32768), I cannot share it with you since its internal

[train_qwen3_8b.py](https://github.com/user-attachments/files/27785229/train_qwen3_8b.py)

**Comment by aroshanghias-nvd** [comment]
I was able to reproduce the reported behavior with a public SQuAD setup, so this does not appear to depend on the customer’s private dataset.

Repro details:
- Commit: `9437ac7c4890a89e61bf8353367093ba4f46dc84`
- Model: `Qwen/Qwen3-0.6B`
- Dataset: SQuAD
- Compared DDP vs Megatron-FSDP with the same training config, changing only `use_megatron_fsdp`

Results:
- DDP job `11878168` completed and converged. Loss went from `6.4113` at iter 1 to roughly `0.3-0.4` by iter 100-150.
- Megatron-FSDP job `11878169` completed with no skipped/nan iterations, but did not converge. Loss started at the same `6.4113` and stayed around `4-5` through iter 200.

Representative lines:
- DDP iter 100: `lm loss: 2.967573E-01`
- DDP iter 150: `lm loss: 3.546824E-01`
- FSDP iter 100: `lm loss: 4.520513E+00`
- FSDP iter 150: `lm loss: 5.203983E+00`
- FSDP iter 200: `lm loss: 4.406222E+00`


Conclusion: confirmed reproducible. Megatron-FSDP completes but shows clear non-convergence compared with DDP on the same public-data repro.

Next step, we will assign this to someone to dive deeper into it.

**Comment by cspades** [comment]
@pavelgein @aroshanghias-nvd Will have to debug this on my side, NeMo 26.04 might be bricked for Megatron-FSDP right now, but what happens if you set `opt_cfg.use_precision_aware_optimizer = False`?

I made multiple fixes recently related to this argument:

- https://github.com/NVIDIA/Megatron-LM/pull/4133 -> Always use the decoupled gradient with precision-aware optimization to avoid spaghetti logic.
- https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3510 -> `self.ddp.megatron_fsdp_use_decoupled_grad = self.optimizer.use_precision_aware_optimizer`
- https://github.com/NVIDIA/Megatron-LM/pull/4427 -> Fixes FusedAdam(use_decoupled_grad=megatron_fsdp and use_precision_aware_optimizer)

Wondering if the 26.04 container release cut off some of these commits. @aroshanghias-nvd Could you share which MCore commit you used inside the MBridge repository that you mounted?

**Comment by aroshanghias-nvd** [comment]
I ran additional SQuAD ablations to isolate this.

Setup:
- MBridge commit: `9437ac7c4890a89e61bf8353367093ba4f46dc84`
- MCore submodule: `815c83d9bcc5854030a456b66f096ae3862292f6`
- Model: `Qwen/Qwen3-0.6B`
- Dataset: SQuAD
- 8 GPUs, `train_iters=200`, `GBS=32`, `MBS=1`

Results:
- DDP baseline converges: iter 100 loss `2.967573E-01`, iter 150 loss `3.546824E-01`
- FSDP with BF16 main params does not converge: iter 100 loss `4.520513E+00`, iter 150 loss `5.203983E+00`, iter 200 loss `4.406222E+00`
- FSDP + `--no-distributed-optimizer`: still does not converge, iter 200 loss `4.406222E+00`
- FSDP + `grad_reduce_in_fp32=True`: still does not converge, iter 200 loss `4.413862E+00`
- FSDP + `megatron_fsdp_main_params_dtype=torch.float32`: converges, iter 100 loss `2.967071E-01`, iter 150 loss `3.541841E-01`, iter 200 loss `1.969704E-01`
- FSDP + `megatron_fsdp_main_params_dtype=torch.float32` + `grad_reduce_in_fp32=True`: also converges, iter 200 loss `1.984223E-01`

Interpretation:
The issue appears specific to Megatron-FSDP with BF16 main params. Disabling distributed optimizer does not fix it, and `grad_reduce_in_fp32=True` alone does not fix it. Setting FSDP main params to FP32 restores DDP-like convergence.

Recommended WAR:

```python
ddp = DistributedDataParallelConfig(
    use_megatron_fsdp=True,
    megatron_fsdp_main_params_dtype=torch.float32,
    ...
)

**Comment by aroshanghias-nvd** [comment]
Meanwhile, @cspades  will further look into it to fix the underlying issue.

**Comment by pavelgein** [comment]
I have run several experiments, based on 26.04 container and Megatron-Bridge from commit 0dae728241d4ead96c5da7aba7d2a383e58bf7d7 

1) Without FSDP (DDP only)
2) With FSDP and  `use_precision_aware_optimizer = True`
3) With FSDP and  `use_precision_aware_optimizer =False` without setting `megatron_fsdp_main_params_dtype`. Fails on gradient clipping
4) With FSDP, `use_precision_aware_optimizer =False` and with setting `megatron_fsdp_main_params_dtype=torch.float32`.

Can confirm that first and fourth runs produce the same results and significantly differ from the second run.

<img width="3711" height="1350" alt="Image" src="https://github.com/user-attachments/assets/71463f58-f66f-4978-a977-fb36e87f1d24" />
